### PR TITLE
feat: add OAuth authentication

### PR DESCRIPTION
**Pull request by SMillerDev**
This _should_ allow API clients to authenticate using OAuth

Relates to #2292

### CODE REVIEW

**Comment by mouse-reeve** [comment]
I thought this was going to be a _way_ longer diff when I went to look at it. I'm going to have to take a look at the library and get a little more familiar, but thank you so much for starting on this!

**Comment by mouse-reeve** [comment]
It looks like this doesn't run yet, so I'm going to mark it as a draft. Feel free to mark it as ready for review once you're ready :)

**Comment by SMillerDev** [comment]
It was just a missing `include`. Should work now.

File: bookwyrm/urls.py [hunk 1]
```diff
@@ -785,6 +785,7 @@
         r"^summary_revoke_key/?$", views.summary_revoke_key, name="summary-revoke-key"
     ),
     path("guided-tour/<tour>", views.toggle_guided_tour),
+    re_path(r'^o/', include('oauth2_provider.urls', namespace='oauth2_provider')),
```
Comment (SMillerDev) [inline bookwyrm/urls.py:788]:
```suggestion
    re_path(r"^o/", include("oauth2_provider.urls", namespace="oauth2_provider")),
```

**Comment by SMillerDev** [comment]
Should be good for another CI run @mouse-reeve

**Comment by rabc** [comment]
Any idea when it can be merged? Then we can move forward with the API implementation :)

**Comment by SMillerDev** [comment]
CI seems okay with it, and I'm still interested in working on this when it's merged. Anyone up for a review ?

**Comment by mouse-reeve** [comment]
Ah I apologize for leaving you hanging on this! I would like to get feedback from other folks on @bookwyrm-social/code-review if yall feel like this is the right first step.

**Comment by Minnozz** [comment]
I think it is currently not very clear what the scope of this PR is. Is it a first step in developing a client API? Can end users already use it after this is merged? If so, should we not integrate OAuth settings into the user preferences, or at least add a link to the management view of the OAuth application? Should there be a test that at least checks if OAuth is working?

**Comment by SMillerDev** [comment]
The goal is to add the Oauth library so we can build on that. After this I plan to:
- add Oauth for users so it can be configured for login
- add an API for user information with that login
- add other APIs

But I'm waiting with the actual configuration because if people don't like this library addition it'll be a lot of reworking

**Comment by SMillerDev** [comment]
So, what can I do to make this mergable?

**Comment by SMillerDev** [comment]
Small prefix to my plans, because of the federated nature I guess I'll need to add something like https://docs.joinmastodon.org/methods/apps/ first so apps can register with the server.

**Comment by mouse-reeve** [comment]
Thank you for your patience with this @SMillerDev -- I've been a little unsure on how to best move forward on this incrementally but I think this is the right first step. Much appreciated!

**Comment by SMillerDev** [comment]
Thanks @mouse-reeve ! Hoping to work on app registration soon

**Comment by SMillerDev** [comment]
https://github.com/bookwyrm-social/bookwyrm/pull/3361 is a first draft of that.

**Pull request merged**